### PR TITLE
⚠️ Refactor golangci-lint config to remove false negatives

### PR DESCRIPTION
**Pull request by vincepri**
**What this PR does / why we need it**:

During some audit of golangci-lint, it was found that one of the exclude
directives was matching a lot more than it should have.

The removal of the unwanted regex caused a large backlog of linting
errors to show up. This changeset brings in a refactor in the
golangci-lint configuration to make sure we catch more errors in the
linting process and enable more linters over time.

Marking this as an ⚠️ breaking change given that in some cases we're unexporting some variables that are only used internally.

/milestone v0.4
/kind release-blocking

### CODE REVIEW

**Comment by vincepri** [comment]
```
sigs.k8s.io/cluster-api/test/framework
  Compatible changes:
  - PodListCondition: added
sigs.k8s.io/cluster-api/cmd/clusterctl/client/repository
  Incompatible changes:
  - NewComponents: changed from func(ComponentsInput) (*components, error) to func(ComponentsInput) (Components, error)
  - NewTemplate: changed from func(TemplateInput) (*template, error) to func(TemplateInput) (Template, error)
sigs.k8s.io/cluster-api/cmd/clusterctl/client/config
  Incompatible changes:
  - InjectConfigPaths: removed
sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha4
  Incompatible changes:
  - ConflictingFileSourceMsg: removed
  - MissingSecretKeyMsg: removed
  - MissingSecretNameMsg: removed
  - PathConflictMsg: removed
sigs.k8s.io/cluster-api/errors
  Incompatible changes:
  - ErrNotImplemented: removed
sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster
  Incompatible changes:
  - Scheme: removed
sigs.k8s.io/cluster-api/test/framework/bootstrap
  Compatible changes:
  - KindClusterProvider: added
sigs.k8s.io/cluster-api/controllers/remote
  Incompatible changes:
  - ClusterCacheControllerName: removed
sigs.k8s.io/cluster-api/exp/controllers
  Incompatible changes:
  - ErrNoAvailableNodes: removed
```

**Comment by fabriziopandini** [comment]
That's definitely a big PT, but I'm glad we are doing this now vs later in the cycle
/lgtm

File: .golangci.yml [hunk 1]
```diff
@@ -43,20 +42,44 @@ issues:
...
   exclude:
-    - Using the variable on range scope `(tc)|(rt)|(tt)|(test)|(testcase)|(testCase)` in function literal
-    - "G108: Profiling endpoint is automatically exposed on /debug/pprof"
-    - Error return value of .((os\.)?std(out|err)\..*|.*Close|.*Flush|os\.Remove(All)?|.*print(f|ln)?|os\.(Un)?Setenv). is not checked
-    # The following are being worked on to remove their exclusion. This list should be reduced or go away all together over time.
-    # If it is decided they will not be addressed they should be moved above this comment.
-    - Subprocess launch(ed with variable|ing should be audited)
-    - (Expect directory permissions to be 0750 or less|Expect file permissions to be 0600 or less)
-    - (G104|G307)
+  - "G108: Profiling endpoint is automatically exposed on /debug/pprof"
```
Comment (sbueringer) [inline .golangci.yml:45]: Would it make sense to migrate those also to exclude-rules to restrict their scope to a specific linter?

File: .golangci.yml [hunk 2]
```diff
@@ -23,11 +23,10 @@ linters:
   - nolintlint
   - prealloc
   - rowserrcheck
-  - scopelint
+  - exportloopref
```
Comment (sbueringer) [inline .golangci.yml:26]: I think the linters are sorted alphabetically

**Comment by vincepri** [comment]
@fabriziopandini @CecileRobertMichon Could you folks double check the incompatible changes from above?

**Comment by sbueringer** [comment]
Apart from the two nits => lgtm

File: .golangci.yml (hunk 2 above)
Comment (vincepri) [reply .golangci.yml:26]: Do we have a linter for that linter 🙃 ? I got no errors locally, but I'd rather not enforce something that cannot be tested/validated

File: .golangci.yml (hunk 1 above)
Comment (vincepri) [reply .golangci.yml:45]: Possibly yeah, should we do that in a follow-up PR?

**Comment by fabriziopandini** [comment]
> @fabriziopandini @CecileRobertMichon Could you folks double check the incompatible changes from above?

@vincepri I double checked incompatible changes and all of them seems reasonable to me

File: .golangci.yml (hunk 2 above)
Comment (sbueringer) [reply .golangci.yml:26]: Nope, but I guess not every review comment can be enforced by linter and the list would be nicer to navigate if it stays sorted :)

File: .golangci.yml (hunk 1 above)
Comment (sbueringer) [reply .golangci.yml:45]: I'll put it as "task" in #4622

File: .golangci.yml (hunk 2 above)
Comment (vincepri) [reply .golangci.yml]: Fixed it, although we should either automate that or just keep it random

**Review by fabriziopandini — COMMENTED** [review]
/lgtm

**Comment by sbueringer** [comment]
@vincepri ex > er but close enough :)

/lgtm

**Pull request merged**